pkg/scalers/kafka_scaler.go

```diff
@@ -430,17 +436,31 @@ func (s *kafkaScaler) getTopicOffsets(partitions []int32) (map[int32]int64, erro
 		request.AddBlock(s.metadata.topic, partitionID, sarama.OffsetNewest, 1)
 	}
 
-	offsets := make(map[int32]int64)
-
 	// Step 2: send requests, one per broker, and collect offsets
+	resultCh := make(chan brokerOffsetResult, len(requests))
+	var wg sync.WaitGroup
+	wg.Add(len(requests))
 	for broker, request := range requests {
-		response, err := broker.GetAvailableOffsets(request)
+		go func(brCopy *sarama.Broker, reqCopy *sarama.OffsetRequest) {
+			response, err := brCopy.GetAvailableOffsets(reqCopy)
+			resultCh <- brokerOffsetResult{response, err}
+			wg.Done()
+		}(broker, request)
+	}
 
-		if err != nil {
-			return nil, err
+	go func() {
+		wg.Wait()
+		close(resultCh)
+	}()
```

Comment:
Should this be inside another goroutine? I mean, aren't we interested to wait all goroutines before continue?
Could we lose any `brokerOffsetResult` in case of don't wait synchronously?

Reply:
Ideally `wg.Wait` should be done in the main goroutine;

Right now the code will still behave correctly because the  `for brokerOffsetRes := range resultCh`.
the range channel will wait for till the channel is closed.

but agree putting wait/close in main goroutine makes it more readable.

ref: https://gobyexample.com/range-over-channels

---
small demo 

```
// You can edit this code!
// Click here and start typing.
package main

import (
	"fmt"
	"sync"
)

func main() {
	wg := sync.WaitGroup{}

	ch := make(chan int, 666)

	for i := 0; i < 10; i++ {
		wg.Add(1)
		go func(i int) {
			ch <- i
			wg.Done()
		}(i)
	}

	go func() {
		wg.Wait()
		close(ch)
	}()

	for out := range ch {
		fmt.Println(out)
	}
}

```

https://go.dev/play/p/pqhlHfu8HtG

Reply:
Indeed @PaulLiang1 . 

In the happy path, when no requests fail, the main goroutine processes all responses correctly and stops when the wg.Wait()ing goroutine closes the channel. 

In the error path, the main goroutine returns on the first error. The remaining requesting goroutines continue their work by putting their results on the buffered channel, but nothing is done with those results. The wait and close  goroutine is useless in this case, as the go GC would just clean up the channel anyways [(ref)](https://stackoverflow.com/questions/8593645/is-it-ok-to-leave-a-channel-open).

I'm gonna try and incorporate the wait and close in the main goroutine :+1:

Reply:
I don't see a way of putting the wg.wait() in the main goroutine. However, an alternative would be to just count the number of response received in the main goroutine and close upon handling all of them. With this approach, the waitingGroup and the extra goroutine are not required anymore. See the snippet below :-). 

```go
	respsReceived := 0
        for resp := range respChan {
              doSomeSyncWork(resp)
              respsReceived++
              if respsReceived == len(requests) {
                      close(respChan)
              }     
        }
```

Reply:
I thought that something like this is enough to wait in main gorouting and it works (but I could be wrong):
```suggestion
	wg.Wait()
	close(resultCh)
```

Reply:
Oh I had the the flow in mind that the main goroutine handles the messages as they come in. WIth your proposed approach, it would only handle the responses once all of the results are in. Since the postprocessing is simple, non-I/O related code I do agree that just waiting until everything in is more than enough :+1:. Thanks for the feedback!

Reply:
and again, I'm not an expert in golang concurrency, so please, take my suggestions like suggestions only xD

Reply:
Me neither, first time playing with golang concurrency :-). Let's get this merged

Reply:
> Oh I had the the flow in mind that the main goroutine handles the messages as they come in. WIth your proposed approach, it would only handle the responses once all of the results are in, and thus, a return on an error can not be returned upon early. However, since the postprocessing is simple, non-I/O related code, and errors should be rare, I do agree that just waiting until all requests are done is enough and easier to understand 👍. WDYT?

what @JorTurFer proposed will work, and:
- Current approach of having `wg.Done` in separate goroutine is closing the channel after all request had finish, so there is no early return anyway;
- To achieve early cancel, there are lib to do that -> [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup) [ref](https://medium.com/swlh/managing-groups-of-gorutines-in-go-ee7523e3eaca)
- but the catch of using errGroup -> need to supply a context to allow cancel propagation -> unfortunately Sarama didn't expose that -> no `ctx.Context`  https://github.com/Shopify/sarama/blob/461b931067e5a03a59a4454ca794a8e5bf83687c/broker.go#L333-L343
- so looks like there is no easy way to achieve early error return